internal/exec/exec.go

```diff
@@ -111,7 +127,13 @@ func (r *Request) execSelections(ctx context.Context, sels []selected.Selection,
 		out.WriteString(f.field.Alias)
 		out.WriteByte('"')
 		out.WriteByte(':')
+		// if this field hasn't finished yet, then it's timed out. Record it as null
+		if !f.finished {
+			out.WriteString("null")
```

Comment:
So here's something that makes me a bit nervous _theoretically_ though practically I think we can get around it. [The spec](http://spec.graphql.org/June2018/#sec-Type-System.Non-Null) denotes that when you add the `!` after a type, the output is non-nullable. If you look around some of our resolver methods, you'll see some return `*TYPE` and some just `TYPE`. 

It seems like our default behaviour is to return null if the context deadline is exceeded. I actually think this makes sense however I think it means this API no longer conforms to the GraphQL spec. We'd need to think of some solution to this.

Reply:
At least internally, I think this would break marshaling logic.

Reply:
I'm seeing a related comment [above](https://github.com/graph-gophers/graphql-go/pull/433/files#diff-87ea4cfcdd6bb8062908727b36ef05d305ddfbd5ed89322e2276f315ac63dd1bR115). 

I believe returning a `null` for a non-nullable field is actually prevented at run time. The schema parser will verify that matching resolver methods have the proper return types. Maybe something like that can be done here? I don't know.

Reply:
The third test I wrote TestTotalFailureOnNonNull shows that the current code will happily record null on non-null features if a timeout has occurred. It doesn't look like the schema parser stops this from happening.

Reply:
> The third test I wrote TestTotalFailureOnNonNull shows that the current code will happily record null on non-null features if a timeout has occurred. It doesn't look like the schema parser stops this from happening.

On master, if you try to do this you'll get errors in your console when you try to query the graphql API so we probably don't want to introduce this unexpected behaviour.

Reply:
I have changed it so the timeout case obeys the non-null behavior we want.

Reply:
Shouldn't the `null` propagate to the nearest null-able parent if a resolver times out? We had this issue before with other errors and I believe that it was fixed, right? So timeout errors should be like any other errors.

Reply:
The issue I was trying to solve here was that the field would not be set to null nor have an error attached to it if a context timeout occurred. We'd start marshaling our result before that resolver ever finished, that's why I am keying off of !finished. I tried to handle the case of propegating that error to the parent above here, this is the case where the field is nullable and a timeout occured.